### PR TITLE
METRON-1695: Expose pcap properties through Ambari

### PR DESCRIPTION
**Pull request by anandsubbu**
## Contributor Comments
This change makes the PCAP config properties available via the Metron service -> Configs tab in Ambari. It also takes care of updating the `pcap.properties` file with relevant values depending on the deploy environment (single node vs. multi node vs. kerberized cluster). 

Here's a screenshot of the PCAP Config:
pcap-config-values

### Testing Steps
* Launch full dev
* Metron PCAP service should come up successfully.
* Go to Ambari -> Metron services -> Configs -> PCAP tab; change any config parameter; 
* Restart the required services and the config change take effect.

### Testing Done
I ran the following different tests to validate the fix.

#### A. Full Dev
* Verified Metron PCAP service comes up successfully 
* Changed some of the PCAP config values from the Ambari UI, validated that this is reflected in the `pcap.properties` file upon service restart.
- **Note**: I ran into [METRON-1786](https://issues.apache.org/jira/browse/METRON-1786) during testing. I was able to resolve the issue by applying the fix in PR #1206 

#### B. Multi-node Cluster
* Validated that the `kafka.zk` property in `pcap.properties` is updated with the correct Broker details
* Changed a PCAP service-related property (E.g. `kafka_pcap_numpackets`) and validated that the `pcap.properties` file is updated.
* Changed a PCAP REST-related property (E.g. `pcap_page_size`) and validated that the `/etc/default/metron` file is updated.

#### C. Kerberized Cluster
* Kerberized the multi-node cluster and setup pycapa
* Validated that `topology.worker.childopts` and `topology.auto-credentials` are 
* Ingested PCAP into the cluster
* Ran queries through the PCAP panel and validated that the modified config settings are in effect (E.g. `pcap_page_size` set to 50). 

## Pull Request Checklist

Thank you for submitting a contribution to Apache Metron.  
Please refer to our [Development Guidelines](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=61332235) for the complete guide to follow for contributions.  
Please refer also to our [Build Verification Guidelines](https://cwiki.apache.org/confluence/display/METRON/Verifying+Builds?show-miniview) for complete smoke testing guides.  

In order to streamline the review of the contribution we ask you follow these guidelines and ask you to double check the following:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? If not one needs to be created at [Metron Jira](https://issues.apache.org/jira/browse/METRON/?selectedTab=com.atlassian.jira.jira-projects-plugin:summary-panel).
- [x] Does your PR title start with METRON-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.
- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

### For code changes:
- [x] Have you included steps to reproduce the behavior or problem that is being changed or addressed?
- [x] Have you included steps or a guide to how the change may be verified and tested manually?
- [NA] Have you ensured that the full suite of tests and checks have been executed in the root metron folder via:
  ```
  mvn -q clean integration-test install && dev-utilities/build-utils/verify_licenses.sh 
  ```

- [ ] Have you written or updated unit tests and or integration tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] Have you verified the basic functionality of the build by building and running locally with Vagrant full-dev environment or the equivalent?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered by building and verifying the site-book? If not then run the following commands and the verify changes via `site-book/target/site/index.html`:

  ```
  cd site-book
  mvn site
  ```

#### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.
It is also recommended that [travis-ci](https://travis-ci.org) is set up for your personal repository such that your branches are built there before submitting a pull request.

### CODE REVIEW

File: metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/package/scripts/params/params_linux.py [hunk 1]
```diff
@@ -387,18 +388,32 @@
 bolt_hdfs_rotation_policy_units = config['configurations']['metron-indexing-env']['bolt_hdfs_rotation_policy_units']
 bolt_hdfs_rotation_policy_count = config['configurations']['metron-indexing-env']['bolt_hdfs_rotation_policy_count']
 
-# Pcap
+# PCAP
 metron_pcap_topology = 'pcap'
 pcap_input_topic = 'pcap'
```
Comment (MohanDV) [inline metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/package/scripts/params/params_linux.py]: pcap input topic is also exposed , so shouldn't we be getting the value from config tab ?

Comment (MohanDV) [reply metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/package/scripts/params/params_linux.py]: pcap input topic is also hardcoded in [status_param.py ](https://github.com/MohanDV/metron/blob/a46fda45b46a9f317e2ef7c3755ce6f072ed893b/metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/package/scripts/params/status_params.py#L123)  we should get the value from config tab here too.

**Comment by nickwallen** [comment]
This is causing one problem on Full Dev. We have 6 slots in Full Dev, but since adding the Pcap Topology, we now require 7 slots for all of the topologies to run.  

What ends up happening is that the 'random_access_indexing' often gets starved.  When spinning up Full Dev, no alerts are visible in the Alerts UI.  We want alerts to be immediately visible as this is a standard validation step.  

We need to figure out a solution around this, which I am not sure what that looks-like right now.  Add a slot?  Do not immediately start 'pcap' topology?  Other?

**Comment by ottobackwards** [comment]
We have an overall deficiency here I believe.  The Slot management is a oft forgotten and manual task, for adding new parsers from the ui etc, you ( I believe ) still have to remember to have already added a slot before starting your parser topology for example.  We may want to consider an overall underlying approach to this that is usable where we fall short.

**Comment by nickwallen** [comment]
> @ottobackwards The Slot management is a oft forgotten and manual task, for adding new parsers from the ...

Sure, worthy of a discuss.  

@anandsubbu I would just drop "snort" as a default for the parser configuration in Ambari.  Then we save a slot and get around this issue for now.  As a follow-on, we could explore other alternatives.

**Comment by mmiklavc** [comment]
Might be worth looking at how we set what spins up by default. For instance, we might not want Pcap to start  by default in *any* installation, let alone full dev. Is there an option in the mpack config for one, or two, is there a way to disable it via our blueprints? This way we don't alter the existing set of topologies. One reason for this is that it's at least mildly useful to have 2+ sensors worth of data writing out to ES that can be viewed in Kibana. And while pcap doesn't go to indexing anyhow, we don't have anything generating data into the pcap topology currently at all. It would just be taking up a slot and doing nothing.

**Comment by nickwallen** [comment]
@anandsubbu I am also running into a separate issue.  After some period of time Ambari will start reporting the status of the pcap topology as stopped, even though it remains running in Storm.  Are you able to replicate this?  I am digging through the logs trying to find some sort of stack trace, but have not found anything yet.

**Comment by ottobackwards** [comment]
@mmiklavc I agree with that, but I also think that we should separate the default sensors from the regular metron install, and have them be their own optional thing.  IE> you would choose to install the snort , yaf, bro things we install automatically now.

**Comment by anandsubbu** [comment]
Latest commit modifies full dev to start only the bro parser topology by default. The PCAP topology now gets the worker slot which was freed up. Note that on a regular deploy, all of bro, snort and yaf topologies will be deployed by default. 

I am also looking for the issue @nickwallen mentioned about the PCAP topology appearing stopped. Nothing so far. Will continue to monitor.

**Comment by MohanDV** [comment]
PCAP related properties needs to be taken out from the [rest_application](https://github.com/apache/metron/blob/master/metron-interface/metron-rest/src/main/config/rest_application.yml#L57). Also the dependency of PCAP on rest settings to be removes as Changing a REST Setting triggers Ambari to [Restart](https://hortonworks.jira.com/browse/BUG-111762) PCAP Topology .

**Comment by anandsubbu** [comment]
@nickwallen suggested a great idea of making use of parser aggregation and thus starting the default parsers on full-dev as an aggregated parser topology. The latest commit makes use of PR #1215 which provides that ability.

Now, on full-dev we start a single aggregated topology for the default configured sensors viz. "bro,snort,yaf".

## Testing Done
- Spun up full-dev and noticed that a single topology named `bro__snort__yaf` is started
- Verified Alerts UI has data from all three sensors:

image

File: metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/package/scripts/params/status_params.py [hunk 2]
```diff
@@ -119,8 +119,8 @@
 metron_keytab_path = config['configurations']['metron-env']['metron_service_keytab']
 
 # Pcap
-metron_pcap_topology = 'pcap'
-pcap_input_topic = 'pcap'
+metron_pcap_topology = config['configurations']['metron-pcap-env']['spout_kafka_topic_pcap']
```
Comment (nickwallen) [inline metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/package/scripts/params/status_params.py]: Quick fix needed.
```
metron_pcap_topology = 'pcap'
```

Comment (anandsubbu) [reply metron-deployment/packaging/ambari/metron-mpack/src/main/resources/common-services/METRON/CURRENT/package/scripts/params/status_params.py]: Done, changed.

**Comment by nickwallen** [comment]
+1 Thanks @anandsubbu !

**Comment by anandsubbu** [comment]
Thanks much @nickwallen for giving it another spin!

Hey @MohanDV , let me know if my below explanation makes sense. @nickwallen , would you have any thoughts on this?

> @MohanDV , as a part of my commit, I have moved all the UI related elements from REST to the PCAP config tab.
> 
> In my understanding, the following parameters are not directly related to the PCAP core service, but are functionally related to REST (used by the PCAP query panel UI):
> 
> * pdml.script.path
> * base.path
> * base.interim.result.path
> * final.output.path
> * page.size
> * yarn.queue
> * finalizer.threadpool.size
> 
> For this reason, I did not move these properties out of the rest_application.yml.
> 
> > @MohanDV Changing a REST Setting triggers Ambari to Restart PCAP Topology
> 
> With my current changes, modifying any config setting in the PCAP tab will require restarting both Metron PCAP and Metron REST services--which is necessary since user can change a PCAP-service related property or a PCAP-REST related property (or both). The converse, which you noted, i.e. changing any REST config does not require restarting the PCAP topology. Here's a screenshot after making changes to REST config...
> 
> image

**Comment by anandsubbu** [comment]
Cool, thanks for your opinion @nickwallen . I will wait for @MohanDV 's thoughts as well before merging.

**Comment by MohanDV** [comment]
> Cool, thanks for your opinion @nickwallen . I will wait for @MohanDV 's thoughts as well before merging.

+1 Thanks @anandsubbu

**Comment by anandsubbu** [comment]
Thanks @MohanDV !

**Pull request closed without merging**